From 8f32f8ea4c4b70b31b15cbe21ab9890108cd4765 Mon Sep 17 00:00:00 2001
From: Jonathan Daugherty <cygnus@foobox.com>
Date: Sun, 4 Nov 2018 07:35:11 -0800
Subject: [PATCH] PictureToSpans: replace specialized lens operators with more
 fundamental ones

This change removes -~ and +~ definitions that can conflict with those
provided by Microlens, and which were just copies of what Microlens now
provides. Instead of using those, this change just inlines their
functionality in the right spots so that we don't have to conditionally
define or import those operators depending on the version of Microlens,
which just recently started to provide these.

This change also goes ahead and introduces an upper bound on Microlens
to avoid future issues like this.
---
 src/Graphics/Vty/PictureToSpans.hs | 14 ++++----------
 vty.cabal                          |  2 +-
 2 files changed, 5 insertions(+), 11 deletions(-)

diff --git a/src/Graphics/Vty/PictureToSpans.hs b/src/Graphics/Vty/PictureToSpans.hs
index b04a82d..9a8bb59 100644
--- a/src/Graphics/Vty/PictureToSpans.hs
+++ b/src/Graphics/Vty/PictureToSpans.hs
@@ -298,11 +298,11 @@ addMaybeClippedJoin name skip remaining offset i0Dim i0 i1 size = do
     state <- get
     when (state^.remaining <= 0) $ fail $ name ++ " with remaining <= 0"
     case state^.skip of
-        s | s > size -> put $ state & skip -~ size
+        s | s > size -> put $ state & skip %~ subtract size
           | s == 0    -> if state^.remaining > i0Dim
                             then do
                                 addMaybeClipped i0
-                                put $ state & offset +~ i0Dim & remaining -~ i0Dim
+                                put $ state & offset %~ (+ i0Dim) & remaining %~ subtract i0Dim
                                 addMaybeClipped i1
                             else addMaybeClipped i0
           | s < i0Dim  ->
@@ -311,10 +311,10 @@ addMaybeClippedJoin name skip remaining offset i0Dim i0 i1 size = do
                     then addMaybeClipped i0
                     else do
                         addMaybeClipped i0
-                        put $ state & offset +~ i0Dim' & remaining -~ i0Dim' & skip .~ 0
+                        put $ state & offset %~ (+ i0Dim') & remaining %~ subtract i0Dim' & skip .~ 0
                         addMaybeClipped i1
           | s >= i0Dim -> do
-                put $ state & skip -~ i0Dim
+                put $ state & skip %~ subtract i0Dim
                 addMaybeClipped i1
         _ -> fail $ name ++ " has unhandled skip class"
 
@@ -346,9 +346,3 @@ snocOp !op !row = do
         when (spanOpsEffectedColumns ops' > regionWidth theRegion)
              $ fail $ "row " ++ show row ++ " now exceeds region width"
         MVector.write theMrowOps row ops'
-
-(-~) :: Num a => ASetter s t a a -> a -> s -> t
-l -~ n = over l (subtract n)
-
-(+~) :: Num a => ASetter s t a a -> a -> s -> t
-l +~ n = over l (n +)
diff --git a/vty.cabal b/vty.cabal
index 4018fed..5b95ea8 100644
--- a/vty.cabal
+++ b/vty.cabal
@@ -45,7 +45,7 @@ library
                        deepseq >= 1.1 && < 1.5,
                        directory,
                        filepath >= 1.0 && < 2.0,
-                       microlens,
+                       microlens < 0.4.11,
                        microlens-mtl,
                        microlens-th,
                        -- required for nice installation with yi
-- 
2.19.1

